File: C++/TwoPointer/hasCycle.cpp
/*
Solution: 双指针，时间复杂度O(n),空间复杂度O(1)
- 定义两个指针slow，fast，从head开始，一个走一步，一个走两步，如果有环一定相遇；
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == NULL)
            return false;
        ListNode *slow = head, *fast = head;
        while(fast->next && fast->next->next){
            slow = slow->next;
            fast = fast->next->next;
            if(fast == slow)
                return true;
        }
        return false;
    }
};

File: Leetcode/Python3/LinkedList/hasCycle.py
# -*- coding:UTF-8 -*-
'''
Solution:
- 定义两个指针，slow，fast
- 一个一次走一步，一个一次走两步，如果相遇，说明有环；
'''
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head or not head.next:
            return False
        
        slow, fast = head, head.next
        while fast and fast.next:
            if slow == fast:
                return True
            slow = slow.next
            fast = fast.next.next
            
        return False
